### ISSUE TITLE
Can't setState inside `hover` 

### ISSUE THREAD

**Issue by eguneys** [gaearon/react-dnd#265]
```
hover (props, monitor, component) {
  const item = monitor.getItem();
  const [left, top] = getAndCalculateDropPosition(monitor);

  component.setState({
    hoverOn: {
      left: left,
      top: top
    }
  });
}
```

I want to use `setState` within hover but that leads to unexpected behavior, what can I do?

**Comment by eguneys** [comment]
The sortable example is using `setState` inside hover, http://gaearon.github.io/react-dnd/examples-sortable-simple.html

Reproduce: 
Drag an item out of drop target, see the `can't drop` icon.
Drag item to another spot to sort, drag item out of drop target, `can't drop` icon is not shown. it is confused.

**Comment by mvestergaard** [comment]
That example doesn't call setState inside hover, it calls a method on props. Also there is no cancel functionality in that example which is why the can't drop icon is not shown.

I had to do something similar to what you want to do, i ended up passing in an event emitter via props. I just used a jquery object $({}). Then i trigger events on it in the hover method, and hook up a listener in the constructor of my component, which sets the state based on the info it receives.

**Comment by gaearon** [comment]
>that leads to unexpected behavior, what can I do?

Can you describe it in more detail, or better, have a reproducible example?

**Comment by eguneys** [comment]
@mvestergaard `the method on props` calls setState that further causes a rerender on droptarget so it counts. 

@gaearon I can reproduce it in the example: http://gaearon.github.io/react-dnd/examples-sortable-simple.html.

1. Drag an item to the right, see can't drop icon, that's normal behaviour.
2. Drag an item over the list, see it sorting, drag item to the right, the can't drop icon is not visible. that's unexpected.

The problem with the icon is real because it confuses if I can drop the item or not. I get unexpected drag drop experience because of this issue. 
Note that I don't have this issue if I remove `setState` from within `hover`.

I noticed this problem while writing `canDrop` method. It was always returning true but still I was getting `can't drop icon`.
And it's not just the icon I am getting unexpected behaviour whether the item is dropped or not.

**Comment by mvestergaard** [comment]
@eguneys true, but it sets the state of a parent component, not the drop target itself, which could probably make a difference.

I'm working on adding reordering with drag/drop to a treeview in a project i'm working on. I wanted it to add a line before or after the target component depending on whether the mouse is in the top or bottom half of the target component. I tried using setState on the component, but it did nothing. The only solution I found that works was the one I described above, using an event emitter. Not sure if that's useful for this issue :)
